### PR TITLE
VST: Vulnerability exploitation (CVE-2019-14287 & CVE-2016-0099)

### PR DESCRIPTION
**Pull request by gerbsec**

### CODE REVIEW

File: endpoint/endpoint.go [hunk 1]
```diff
@@ -143,7 +143,7 @@ func Shell(args []string) string {
 	if err != nil {
 		if exitError, ok := err.(*exec.ExitError); ok {
 			println(string(exitError.Stderr))
-			os.Exit(exitError.ExitCode())
+			return fmt.Sprintf("%d", exitError.ExitCode())
```
Comment (bfuzzy1) [inline endpoint/endpoint.go:146]: Check https://github.com/preludeorg/test/blob/be9040495ffa912e10f4bea5a2737e69d29713fb/endpoint/endpoint.go

I added a different way to handle errors. Open to feedback on it but it works for the system utilities PR.

Comment (gerbsec) [reply endpoint/endpoint.go:146]: i like yours better

Comment (bfuzzy1) [reply endpoint/endpoint.go:146]: Ok make any changes that need to be made so I can test please.

**Comment by bfuzzy1** [comment]
@privateducky this depends on the endpoint changes in this PR https://github.com/preludeorg/test/pull/74

File: endpoint/endpoint.go [hunk 2]
```diff
@@ -29,11 +29,11 @@ func Stop(code int) {
 }
 
 func RunWithTimeout(function fn) {
-    go function()
-    select {
-    case <- time.After(10 * time.Second):
-        os.Exit(102)
-    }
+	go function()
```
Comment (privateducky) [inline endpoint/endpoint.go:32]: why is this change in the PR?

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Block exploit attempts - Privilege Escalation
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:1]: title doesn't need to preface itself with the rule name. provide a better name.

File: endpoint/endpoint.go (hunk 2 above)
Comment (gerbsec) [reply endpoint/endpoint.go:32]: it was an automatic syntax fix, we decided to just keep it

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+Exploit attempts represent a significant security threat by allowing attackers to gain unauthorized access, perform malicious actions, and potentially lead to data theft or downtime. Blocking exploit attempts is critical to maintaining system security and preventing negative consequences, as well as protecting against privilege escalation vulnerabilities, which enable attackers to bypass security controls and gain higher levels of access than authorized. 
+
+Example output:
+```
+[+] Starting Test
+[-] Test was not prevented
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:8]: let's add a print line about the exploit type being attempted

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Block exploit attempts - Privilege Escalation
+
+Exploit attempts represent a significant security threat by allowing attackers to gain unauthorized access, perform malicious actions, and potentially lead to data theft or downtime. Blocking exploit attempts is critical to maintaining system security and preventing negative consequences, as well as protecting against privilege escalation vulnerabilities, which enable attackers to bypass security controls and gain higher levels of access than authorized.
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:3]: let's bounce this off the @preludeorg/wizards team for a more descriptive/better description.

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+Example output:
+```
+[+] Starting Test
+[-] Test was not prevented
+```
+Defenses should prevent exploit attempts from running.
+
+## How
+
+> Safey: This VST exploits the vulnerabilities and only runs the command `whoami` as a proof-of-concept.
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:14]: weird phrasing for this sentence; let's do better

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# Block exploit attempts - Privilege Escalation
+
+Exploit attempts represent a significant security threat by allowing attackers to gain unauthorized access, perform malicious actions, and potentially lead to data theft or downtime. Blocking exploit attempts is critical to maintaining system security and preventing negative consequences, as well as protecting against privilege escalation vulnerabilities, which enable attackers to bypass security controls and gain higher levels of access than authorized. 
+
+Example output:
+```
+[+] Starting Test
+[-] Test was not prevented
+```
+Defenses should prevent exploit attempts from running.
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:10]: Change to: "Defenses should recognize an exploit attempt and stop it from completing."

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+## How
+
+> Safey: This VST exploits the vulnerabilities and only runs the command `whoami` as a proof-of-concept.
+
+This VST exploits two different vulnerabilities, depending on the operating system, by targeting CVE-2019-14287 for Linux or Darwin, or CVE-2016-0099 for Windows, allowing a user with limited privileges to escalate their privileges and run `whoami` as a proof-of-concept. Blocking these exploits is necessary to ensure the security and integrity of a system, preventing potential attacks and their negative consequences.
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:16]: does this work on linux/Mac? every test doesn't need to obviously, but as a .ps1 file, it seems unlikely we'd be testing on those machines... not too common open source PowerShell is installed there.. if it does work out of box on those OS', then nothing to change -- just curious.

Comment (privateducky) [reply tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:16]: The second sentence of this description has nothing to do with explaining the test. We need this whole block to be rewritten, describing exactly how the test works.

Is this test being attributed anywhere?

Comment (gerbsec) [reply tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:16]: I have 1 linux/mac priv esc exploit coupled with 1 ps1 windows exploit.

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/d36c6715-ac9d-4d22-b6d3-cc778dac473b.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
+/*
+ID: d36c6715-ac9d-4d22-b6d3-cc778dac473b
+NAME: Can a low priv user privilege escalate to an administrator?
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/d36c6715-ac9d-4d22-b6d3-cc778dac473b.go:3]: this should match the README name exactly.

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md (hunk 5 above)
Comment (privateducky) [reply tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:3]: this description needs to fully describe the CVEs under test here; we can pull from the NVD descriptions and attribute the text.

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/d36c6715-ac9d-4d22-b6d3-cc778dac473b.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+)
+
+//go:embed Invoke-MS16032.ps1
+var ms16032 []byte
+var dir, _ = os.Getwd()
+var path = filepath.Join(dir, "Invoke-MS16032.ps1")
+
+var supported = map[string][]string{
+	"windows": {"powershell.exe", "-c", path},
+	"darwin":  {"bash", "-c", "sudo -u#-1 id"},
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/d36c6715-ac9d-4d22-b6d3-cc778dac473b.go:24]: does one need to run this command as root/sudo to get the answer?

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/d36c6715-ac9d-4d22-b6d3-cc778dac473b.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+func test() {
+	command := supported[runtime.GOOS]
+	if runtime.GOOS == "windows" {
+		Endpoint.Write(path, ms16032)
+	}
+	cmd, err := Endpoint.Shell(command)
+	if err != nil {
+		println("[+] The test was prevented or machine not vulnerable")
+		Endpoint.Stop(107)
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/d36c6715-ac9d-4d22-b6d3-cc778dac473b.go:36]: let's add this code to the docs

**Review by privateducky — CHANGES_REQUESTED** [review]
several things to adjust

**Comment by bfuzzy1** [comment]
This is pending README recommendations from @preludeorg/wizards .

**Comment by bfuzzy1** [comment]
@gerbsec test this with the shell updates please

**Review by bfuzzy1 — CHANGES_REQUESTED** [review]
Retest.

**Comment by gerbsec** [comment]
> Retest.

Tested! GTG!

**Review by bfuzzy1 — APPROVED** [review]

**Comment by bfuzzy1** [comment]
> > Retest.
> 
> Tested! GTG!

Thank you. Approved.

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+[+] Testing CVE-2019-14287 on Linux/Darwin OR CVE-2016-0099 on Windows.
+[-] Test failed (unauthorized access to root was gained)
+```
+Properly updated systems should not allow a non-root user to run unauthorized commands as root.
+
+## How
+
+> Safety: As a proof-of-concept, this VST utilizes the identified vulnerabilities and only executes the "whoami" command
+
+This test involves running the "whoami" command as a non-root user, which outputs the user ID of the user who runs it and can be used to determine if the machine is vulnerable to the CVEs. This test is safe because running this command makes no lasting change (it only outputs a text value), and no other commands are run.
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:19]: "is vulnerable to the mapped CVEs."

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Can an arbitrary (non-root) user execute commands as root?
```
Comment (privateducky) [inline tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md:1]: change to: "Can a regular user run administrator-level commands?"

**Review by privateducky — CHANGES_REQUESTED** [review]
two small items

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md (hunk 12 above)
Comment (bfuzzy1) [reply tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md]: done.

File: tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md (hunk 13 above)
Comment (bfuzzy1) [reply tests/d36c6715-ac9d-4d22-b6d3-cc778dac473b/README.md]: done.

**Review by privateducky — APPROVED** [review]
when this is uploaded, please verify it gets mapped to the appropriate CVEs. @packplusplus knows how to do this.

**Comment by packplusplus** [comment]
Done in staging. I'll upload it to prod when we get the next service release out. 

```
{
  "account_id": "prelude",
  "id": "d36c6715-ac9d-4d22-b6d3-cc778dac473b",
  "name": "Can a regular user run administrator-level commands?",
  "attachments": [
    "Invoke-MS16032.ps1",
    "README.md",
    "d36c6715-ac9d-4d22-b6d3-cc778dac473b.go",
    "d36c6715-ac9d-4d22-b6d3-cc778dac473b_darwin-arm64",
    "d36c6715-ac9d-4d22-b6d3-cc778dac473b_darwin-x86_64",
    "d36c6715-ac9d-4d22-b6d3-cc778dac473b_linux-arm64",
    "d36c6715-ac9d-4d22-b6d3-cc778dac473b_linux-x86_64",
    "d36c6715-ac9d-4d22-b6d3-cc778dac473b_windows-x86_64"
  ],
  "mappings": [
    "CVE-2019-14287",
    "CVE-2016-0099"
  ]
}
```

**Pull request merged**